### PR TITLE
Gui 2366 Implement graphs on metrics landing page

### PR DESCRIPTION
**Pull request by dkavanagh**
https://eucalyptus.atlassian.net/browse/GUI-2366

### CODE REVIEW

**Comment by euca-qa-gui-tester** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.qa1.eucalyptus-systems.com//job/user-console-PR-monitor/1522/
Test FAILed.

**Comment by euca-qa-gui-tester** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.qa1.eucalyptus-systems.com//job/user-console-PR-monitor/1561/
Test FAILed.

**Comment by euca-qa-gui-tester** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.qa1.eucalyptus-systems.com//job/user-console-PR-monitor/1565/
Test FAILed.

**Comment by euca-qa-gui-tester** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.qa1.eucalyptus-systems.com//job/user-console-PR-monitor/1573/
Test FAILed.

**Comment by euca-qa-gui-tester** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.qa1.eucalyptus-systems.com//job/user-console-PR-monitor/1574/
Test FAILed.

**Comment by euca-qa-gui-tester** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.qa1.eucalyptus-systems.com//job/user-console-PR-monitor/1581/
Test FAILed.

**Comment by dkavanagh** [comment]
hal, retest this please

**Comment by euca-qa-gui-tester** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.qa1.eucalyptus-systems.com//job/user-console-PR-monitor/1618/
Test FAILed.

**Comment by dkavanagh** [comment]
hal, retest this please

File: eucaconsole/static/js/pages/cloudwatch_charts.js [hunk 1]
```diff
@@ -13,8 +13,13 @@ angular.module('CloudWatchCharts', ['EucaConsoleUtils', 'ChartAPIModule', 'Chart
 .controller('CloudWatchChartsCtrl', function ($scope, eucaUnescapeJson, eucaOptionsArray) {
     var vm = this;
```
Comment (coyote240) [inline eucaconsole/static/js/pages/cloudwatch_charts.js]: The var 'vm' here is unnecessary.  Because all functions are defined within the scope of the controller method, you can and probably should access them directly.  There is a minor efficiency difference as well, but not enough to matter.

The only time you would need vm in this context would be in the rare occasion that you were returning a closure from a child function, but this is pretty rare in angular.  In that case, one would typically assign 'this' to an auxiliary variable close to where it is needed for clarity.

File: eucaconsole/static/js/pages/cloudwatch_charts.js [hunk 2]
```diff
@@ -127,6 +213,38 @@ angular.module('CloudWatchCharts', ['EucaConsoleUtils', 'ChartAPIModule', 'Chart
...
         });
+        $scope.$watch('dimensions', function(newVal, oldVal) {
+            if (newVal != oldVal) {
+                renderChart($scope);
+            }
+        });
+        if ($scope.large) {
+            var parentCtrl = $scope.$parent.chartsCtrl;
+            // Granularity (period) defaults to 5 min, so no need to pass it here
+            var options = {
```
Comment (coyote240) [inline eucaconsole/static/js/pages/cloudwatch_charts.js]: Not clear to me where 'options' is used in this scope.

File: eucaconsole/static/js/pages/metrics.js [hunk 3]
```diff
@@ -101,12 +120,60 @@ angular.module('MetricsPage', ['LandingPage', 'CloudWatchCharts', 'EucaConsoleUt
             }
             // set sticky table headers
             $('table.table').stickyTableHeaders({scrollableArea: $(".split-top")});
+            // parse graph pre-selection
+            params = $.url().param();
+            if (params.graph !== undefined) {
+                // parse graph params
+                var graph = purl("?"+$.base64.decode(params.graph)).param();
+                console.log("chart from URL = "+$.base64.decode(params.graph));
```
Comment (coyote240) [inline eucaconsole/static/js/pages/metrics.js]: Probably want to drop the console.log

File: eucaconsole/static/js/pages/cloudwatch_charts.js (hunk 1 above)
Comment (kamalgill) [inline eucaconsole/static/js/pages/cloudwatch_charts.js]: @coyote240: Just a heads up that this pattern is advocated in the Angular Style Guide at https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md#controlleras-with-vm

Comment (coyote240) [inline eucaconsole/static/js/pages/cloudwatch_charts.js]: Yes, but as a solution to a problem.  It may be noted, even in the examples cited in the article, that the reason vm works is that it is available in the containing scope.  The example provided is also a contrivance in that it would not necessarily be encouraged to $watch variables outside of $scope, or to use variables outside of $scope in an ng-model.
My point is, all of these are unnecessary overhead.  While there are times when it is necessary to capture the value of 'this', they are few, and I would not recommend any of the uses of the 'vm' variable in this article.

Comment (coyote240) [inline eucaconsole/static/js/pages/cloudwatch_charts.js]: The rest of the style guide, however, is pretty good in my opinion.  Some of the layout style I'm not that into, but the motivations behind a lot of the design decisions are pretty good.

**Comment by dkavanagh** [comment]
hal, retest this please

File: eucaconsole/static/js/pages/metrics.js [hunk 4]
```diff
@@ -24,12 +24,31 @@ angular.module('MetricsPage', ['LandingPage', 'CloudWatchCharts', 'EucaConsoleUt
             }
         };
     })
+    .directive('datepicker', function () {
+        return {
+            require: 'ngModel',
+            restrict: 'A',
+            scope: { format: "=" },
```
Comment (dkavanagh) [inline eucaconsole/static/js/pages/metrics.js]: map ngModel here

File: eucaconsole/static/js/pages/metrics.js [hunk 5]
```diff
@@ -24,12 +24,31 @@ angular.module('MetricsPage', ['LandingPage', 'CloudWatchCharts', 'EucaConsoleUt
...
+            link: function(scope, element, attrs, ngModel){
+                if(typeof(scope.format) == "undefined"){ scope.format = "yyyy/mm/dd hh:ii"; }
+                var startDate = new Date();
+                startDate.setHours(-(14 * 24));  // move back 2 weeks
+                var endDate = new Date();
+                $(element).fdatepicker({format: scope.format, pickTime: true, startDate:startDate, endDate:endDate}).on('changeDate', function(ev){
+                    scope.$apply(function(){
+                        ngModel.$setViewValue(ev.date);
```
Comment (dkavanagh) [inline eucaconsole/static/js/pages/metrics.js]: then just assign model value directly

File: eucaconsole/static/js/pages/metrics.js [hunk 6]
```diff
@@ -24,12 +24,31 @@ angular.module('MetricsPage', ['LandingPage', 'CloudWatchCharts', 'EucaConsoleUt
...
+            require: 'ngModel',
+            restrict: 'A',
+            scope: { format: "=" },
+            link: function(scope, element, attrs, ngModel){
+                if(typeof(scope.format) == "undefined"){ scope.format = "yyyy/mm/dd hh:ii"; }
+                var startDate = new Date();
+                startDate.setHours(-(14 * 24));  // move back 2 weeks
+                var endDate = new Date();
+                $(element).fdatepicker({format: scope.format, pickTime: true, startDate:startDate, endDate:endDate}).on('changeDate', function(ev){
+                    scope.$apply(function(){
```
Comment (dkavanagh) [inline eucaconsole/static/js/pages/metrics.js]: may be able to do without $apply.. try it

File: eucaconsole/static/js/pages/cloudwatch_charts.js (hunk 2 above)
Comment (dkavanagh) [inline eucaconsole/static/js/pages/cloudwatch_charts.js]: wow, not finding that bit of code anymore

File: eucaconsole/static/js/pages/metrics.js (hunk 3 above)
Comment (dkavanagh) [inline eucaconsole/static/js/pages/metrics.js]: oh, not this either, but I realized you might have reviewed code that had tvo's commits? Dunno. not finding the console.log stmt.

File: eucaconsole/static/js/pages/metrics.js (hunk 4 above)
Comment (dkavanagh) [inline eucaconsole/static/js/pages/metrics.js]: nor this.. did someone do another commit here?

File: eucaconsole/static/js/pages/cloudwatch_charts.js (hunk 1 above)
Comment (dkavanagh) [inline eucaconsole/static/js/pages/cloudwatch_charts.js]: I can say switching out the vm references for $scope broke a lot of stuff. No time to dig into that, but I think it's related to the "controller as" stuff in the template.

Comment (kamalgill) [inline eucaconsole/static/js/pages/cloudwatch_charts.js]: @dkavanagh: I'd vote to keep the controllerAs stuff around.  It's fine to adhere to Papa's style guide IMO, as it's a bit more forward-looking than reverting to the $scope-based approach.

File: eucaconsole/static/js/thirdparty/nvd3/nv.d3.js [hunk 7]
```diff
@@ -4882,13 +4882,14 @@ nv.models.lineChart = function() {
                     .call(legend);
 
                 if ( margin.top != legend.height()) {
-                    margin.top = legend.height();
-                    availableHeight = (height || parseInt(container.style('height')) || 400)
-                        - margin.top - margin.bottom;
+                    //margin.top = legend.height();
```
Comment (coyote240) [inline eucaconsole/static/js/thirdparty/nvd3/nv.d3.js]: Commented code.

File: eucaconsole/static/js/pages/cloudwatch_charts.js [hunk 8]
```diff
@@ -37,6 +41,26 @@ angular.module('CloudWatchCharts', ['EucaConsoleUtils', 'ChartAPIModule', 'Chart
     vm.specifyZonesMetrics = [  // Pass availability zones for certain metrics
         'HealthyHostCount', 'UnHealthyHostCount'
     ];
+    Date.prototype.datetime = function() {
+        var mm = (this.getMonth()+1).toString();
+        var dd = this.getDate().toString();
+        var hh = this.getHours().toString();
+        var ii = this.getMinutes().toString();
+        return this.getFullYear()+"/"+(mm.length===2?mm:"0"+mm[0])+"/"+(dd.length===2?dd:"0"+dd[0])+" "+(hh.length===2?hh:"0"+hh[0])+":"+(ii.length===2?ii:"0"+ii[0]);
```
Comment (coyote240) [inline eucaconsole/static/js/pages/cloudwatch_charts.js]: Can this be formatted?

Comment (dkavanagh) [inline eucaconsole/static/js/pages/cloudwatch_charts.js]: Oh, sure. I guess I was trying to keep it as fast as possible, so not sure format vs string concat is faster.
